presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/ReorderJoins.java

```diff
@@ -0,0 +1,274 @@
+            PlanNode right = placeholderJoin.getRight();
+            if (left instanceof JoinGraphNode) {
+                left = chooseJoinOrder((JoinGraphNode) left);
+            }
+            if (right instanceof JoinGraphNode) {
+                right = chooseJoinOrder((JoinGraphNode) right);
+            }
+
+            JoinNode joinNode = (JoinNode) placeholderJoin.replaceChildren(ImmutableList.of(left, right));
+            return chooseJoinNodeProperties(joinNode);
+        }
+
+        private JoinNode chooseJoinNodeProperties(JoinNode joinNode)
+        {
+            List<JoinNode> possibleJoinNodes = new ArrayList<>();
+            String joinDistributionType = getJoinDistributionType(session);
+            if (joinDistributionType.equals(FeaturesConfig.JoinDistributionType.REPARTITIONED) || joinDistributionType.equals(FeaturesConfig.JoinDistributionType.AUTOMATIC)) {
+                possibleJoinNodes.add(joinNode.withDistributionType(PARTITIONED));
+                possibleJoinNodes.add(joinNode.flip().withDistributionType(PARTITIONED));
+            }
+            if (joinDistributionType.equals(FeaturesConfig.JoinDistributionType.REPLICATED) || joinDistributionType.equals(FeaturesConfig.JoinDistributionType.AUTOMATIC)) {
+                possibleJoinNodes.add(joinNode.withDistributionType(REPLICATED));
+                possibleJoinNodes.add(joinNode.flip().withDistributionType(REPLICATED));
+            }
+
+            return planNodeOrdering().min(possibleJoinNodes);
+        }
+
+        private Ordering<PlanNode> planNodeOrdering()
+        {
+            return new Ordering<PlanNode>()
+            {
+                @Override
+                public int compare(PlanNode node1, PlanNode node2)
+                {
+                    PlanNodeCostEstimate node1Cost = lookup.getCumulativeCost(node1, session, symbolAllocator.getTypes());
+                    PlanNodeCostEstimate node2Cost = lookup.getCumulativeCost(node2, session, symbolAllocator.getTypes());
+                    return costComparator.compare(session, node1Cost, node2Cost);
+                }
+            };
+        }
+    }
+
+    /**
+     * This method generates all the ways of dividing  totalNodes into two sets
+     * each containing at least one node. It will generate one set for each
+     * possible partitioning. The other partition is implied in the absent values.
+     * In order not to generate the inverse of any set, we always include the 0th
+     * node in our sets.
+     *
+     * @param totalNodes
+     * @return A set of sets each of which defines a partitioning of totalNodes
+     */
+    public static Set<Set<Integer>> generatePartitions(int totalNodes)
+    {
+        Set<Integer> numbers = IntStream.range(0, totalNodes)
+                .boxed()
+                .collect(toImmutableSet());
+        return Sets.powerSet(numbers).stream()
+                .filter(subSet -> subSet.contains(0))
```

Comment:
I think this will be less performant because
- "materializing" power set is pretty cheap
- filtering is pretty cheap too

So my guess is that's actually slower. But I didn't benchmark that. Nor that I think it matters much -- it doesn't improve the algorithmic complexity.